### ISSUE TITLE
The whole node-kubelet-serial job is failing because of global timeout

### ISSUE THREAD

**Issue by cynepco3hahue** [kubernetes/kubernetes#102148]
#### Which jobs are failing:
https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial

#### Which test(s) are failing:

Impossible to say, the lane is failing because of the global timeout and does not have any attached artifacts.

#### Since when has it been failing:
At least for two weeks

#### Testgrid link:
https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial

**Comment by wzshiming** [comment]
It seems to be the impact after this PR https://github.com/kubernetes/test-infra/pull/21828

/cc @odinuge

**Comment by odinuge** [comment]
> @cynepco3hahue: At least for two weeks

It has been failing for months, if not years... :/

> @wzshiming It seems to be the impact after this PR kubernetes/test-infra#21828

It did consistently do this before kubernetes/test-infra#21828 as well, but we no longer see any OOMs now tho. So it might be possible that the tests run fine, and that we only need to increase the timeout to get it properly working. Would be awesome if someone could take a look at it.

**Comment by odinuge** [comment]
If selecting `show stale tests`, you see that the test run sometimes, even though it looks like it times out every time.

https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial&show-stale-tests=

**Comment by odinuge** [comment]
Hmm, yeah, it looks like the apiserver dies because of too many open files, and the e2e suite doesn't restart it...

**Comment by fromanirh** [comment]
@odinuge great catch! :) Now I wonder (and I'll check reviewing the tests) that is intentional for the test to kill the apiserver. I wonder how it is supposed to work in the ad-hoc environment the `e2e_node` tests run into.

**Comment by odinuge** [comment]
> @odinuge great catch! :) Now I wonder (and I'll check reviewing the tests) that is intentional for the test to kill the apiserver. I wonder how it is supposed to work in the ad-hoc environment the e2e_node tests run into.

I know a lot of tests are intentionally killing and restarting kubelet, but I don't think that killing the apiserver is wanted. From the tooling where it is started, it looks like there is no tooling around restarting it;

https://github.com/kubernetes/kubernetes/blob/925900317e43e58435082f624f5969e3cfe25c67/test/e2e_node/services/apiserver.go#L58-L117

**Comment by fromanirh** [comment]
> > @odinuge great catch! :) Now I wonder (and I'll check reviewing the tests) that is intentional for the test to kill the apiserver. I wonder how it is supposed to work in the ad-hoc environment the e2e_node tests run into.
> 
> I know a lot of tests are intentionally killing and restarting kubelet, but I don't think that killing the apiserver is wanted. From the tooling where it is started, it looks like there is no tooling around restarting it;
> 
> https://github.com/kubernetes/kubernetes/blob/925900317e43e58435082f624f5969e3cfe25c67/test/e2e_node/services/apiserver.go#L58-L117

Yep, this is what I was thinking. Restarting/killing kubelet is fine, and supported, we do it in quite some places (at very least cpu, topology manager e2e tests, also memory manager tests if memory serves). It's killing apiserver and/or etcd, and in general everything-but-kubelet which looks suspicious **in the context of the e2e_node tests**.

Maybe these tests are just misplaced? should they be moved in the regular e2e suite, on which they run against a full cluster and not on a mini-, special purpose environment like e2e_node tests are running?

**Comment by odinuge** [comment]
> Yep, this is what I was thinking. Restarting/killing kubelet is fine, and supported, we do it in quite some places (at very least cpu, topology manager e2e tests, also memory manager tests if memory serves). It's killing apiserver and/or etcd, and in general everything-but-kubelet which looks suspicious in the context of the e2e_node tests.

Yeah. I think a simple "restart" apiserver if it dies would work. It would however mask potential performance issues, so I think the best option long term is to look at why the apiserver dies; and try to fix that. 

> Maybe these tests are just misplaced? should they be moved in the regular e2e suite, on which they run against a full cluster and not on a mini-, special purpose environment like e2e_node tests are running?

I think that isn't possible, since the tests in `node-kubelet-serial` need to be serially executed, and they require full access over kubelet and its lifecycle, together with modifying node stuff. They are also pretty disruptive, especially when testing eviction related things.

If the regular e2e tests can allow for this, then I guess that is ok, but I don't think that is the case.

**Comment by fromanirh** [comment]
> > Yep, this is what I was thinking. Restarting/killing kubelet is fine, and supported, we do it in quite some places (at very least cpu, topology manager e2e tests, also memory manager tests if memory serves). It's killing apiserver and/or etcd, and in general everything-but-kubelet which looks suspicious in the context of the e2e_node tests.
> 
> Yeah. I think a simple "restart" apiserver if it dies would work. It would however mask potential performance issues, so I think the best option long term is to look at why the apiserver dies; and try to fix that.


This surely fixes the current issue, and we could do that. But I wonder if it is the right thing to do in the e2e_node flow. I mean: should we handle these restarts in the special purpose reduced environment? are these tests already abusing this environment?
I'll start doing some archeology in git history as soon as I have more time to spend on this issue.

> > Maybe these tests are just misplaced? should they be moved in the regular e2e suite, on which they run against a full cluster and not on a mini-, special purpose environment like e2e_node tests are running?
> 
> I think that isn't possible, since the tests in `node-kubelet-serial` need to be serially executed, and they require full access over kubelet and its lifecycle, together with modifying node stuff. They are also pretty disruptive, especially when testing eviction related things.
> 
> If the regular e2e tests can allow for this, then I guess that is ok, but I don't think that is the case.

Good point. I agree.

**Comment by odinuge** [comment]
> This surely fixes the current issue, and we could do that. But I wonder if it is the right thing to do in the e2e_node flow. I mean: should we handle these restarts in the special purpose reduced environment? are these tests already abusing this environment?
I'll start doing some archeology in git history as soon as I have more time to spend on this issue.

Agree!

I guess we can do something like this: https://github.com/kubernetes/kubernetes/pull/102169 and see if it mitigates the issue. In case it is actually running with `1024`, that isn't a realistic production environment anyways. What do you think about that?

**Comment by fromanirh** [comment]
> > This surely fixes the current issue, and we could do that. But I wonder if it is the right thing to do in the e2e_node flow. I mean: should we handle these restarts in the special purpose reduced environment? are these tests already abusing this environment?
> > I'll start doing some archeology in git history as soon as I have more time to spend on this issue.
> 
> Agree!
> 
> I guess we can do something like this: #102169 and see if it mitigates the issue. In case it is actually running with `1024`, that isn't a realistic production environment anyways. What do you think about that?

This looks like a fine change and we should most likely fix the open files issue anyway. So thanks for sending the PR!

**Comment by odinuge** [comment]
I also found an instance of `E0510 14:49:01.733215    3236 dynamic_serving_content.go:164] key failed with : open /var/run/kubernetes/apiserver.crt: too many open files`, and that doesn't look that good :sweat_smile: 

Would be nice if anyone could take a look at increasing the timeout in test-infra as well, and see if that makes any difference.

**Comment by harche** [comment]
@odinuge https://github.com/kubernetes/test-infra/pull/22290. Still work in progress though.

**Comment by odinuge** [comment]
After some digging, I found out that the reason for the lack of stdout and artifacts from this test suite is the way it is killed upon the timout. When the global timeout is exceeded, SIGINT is sent to all the processes. If we ignore the first SIGINT in node-e2e, we can print logs to stdout and fetch artifacts. In case the test is still running 15 minutes later, it will be killed via SIGKILL.

See #102209. Would be nice to get that merged as soon as possible!

**Comment by fromanirh** [comment]
> After some digging, I found out that the reason for the lack of stdout and artifacts from this test suite is the way it is killed upon the timout. When the global timeout is exceeded, SIGINT is sent to all the processes. If we ignore the first SIGINT in node-e2e, we can print logs to stdout and fetch artifacts. In case the test is still running 15 minutes later, it will be killed via SIGKILL.
> 
> See #102209. Would be nice to get that merged as soon as possible!

this is a good idea, but is seems racy because it seems there is a hidden assumption we have enough time to collect the log and the artifacts. Which is a big step forward anyway! but let me ask: do we have a simple way to, once the process receives SIGINT, start explicitely the process of waiting for childs to end, collect logs + artifacts and then terminate (instead of just ignoring the signal and wait for things to converge implicitely)?
If so maybe worth making it a bit more robust; if not, let's go ahead with this fix anyway/

**Comment by odinuge** [comment]
Thanks for looking @fromanirh!

> this is a good idea, but is seems racy because it seems there is a hidden assumption we have enough time to collect the log and the artifacts.

The executor will wait 15min before sending SIGKILL. That should be ok with time. If that doesn't work, we can increase that later.

> but let me ask: do we have a simple way to, once the process receives SIGINT, start explicitely the process of waiting for childs to end, collect logs + artifacts and then terminate (instead of just ignoring the signal and wait for things to converge implicitely)?

Since the runner will send SIGINT to all children processes, the SSH connections, they will be terminated by SIGINT. Printing of logs and collection of artifacts will therefore start imminently after getting the SIGINT. This worked well when running locally, and _hopefully_ it will work the same way here.

> If so maybe worth making it a bit more robust; if not, let's go ahead with this fix anyway/ 

Yes, at a later stage that might be smart. The big issue here is that we use the same timeout for ginkgo and for the global test runner; resulting in the fact that the global test runner will always kill the test before ginkgo hits its timeout.

Does that make sense?

Tried to describe it here as well: https://github.com/kubernetes/kubernetes/pull/102209/files#diff-68ff9f8d3e1907be30edb5c5467ca28bc054de257dd0a548bbee2d5af382e85aR204-R207

**Comment by fromanirh** [comment]
> Thanks for looking @fromanirh!
> 
> > this is a good idea, but is seems racy because it seems there is a hidden assumption we have enough time to collect the log and the artifacts.
> 
> The executor will wait 15min before sending SIGKILL. That should be ok with time. If that doesn't work, we can increase that later.

The thing that worries me it seems we only rely on time and to a lot of implicit assumptions in general.

> > but let me ask: do we have a simple way to, once the process receives SIGINT, start explicitely the process of waiting for childs to end, collect logs + artifacts and then terminate (instead of just ignoring the signal and wait for things to converge implicitely)?
> 
> Since the runner will send SIGINT to all children processes, the SSH connections, they will be terminated by SIGINT. Printing of logs and collection of artifacts will therefore start imminently after getting the SIGINT. This worked well when running locally, and _hopefully_ it will work the same way here.

I fully ack we are in deep red zone and we need to fix the lane. I'm totally onboard with the urgency (and the sorry state of the lane in general). I think we should not rush fixes in anyway which could make our life harder in the future.

OTOH it seems that better and more robust handling of timeouts and artifacts collection need some more investigation and a bit more work; your change clearly signals what's going on and why we do what we do, so I'm inclined to go ahead with it.

> > If so maybe worth making it a bit more robust; if not, let's go ahead with this fix anyway/
> 
> Yes, at a later stage that might be smart. The big issue here is that we use the same timeout for ginkgo and for the global test runner; resulting in the fact that the global test runner will always kill the test before ginkgo hits its timeout.
> 
> Does that make sense?

Yes, it does make sense to me. Your fix seems to hit the sweet spot between unlocking things without making the overall infra more fragile.
Still, this is one more hint the flow as whole need some attention and some improvements.

> Tried to describe it here as well: https://github.com/kubernetes/kubernetes/pull/102209/files#diff-68ff9f8d3e1907be30edb5c5467ca28bc054de257dd0a548bbee2d5af382e85aR204-R207

Yep, thanks.

**Comment by odinuge** [comment]
> The thing that worries me it seems we only rely on time and to a lot of implicit assumptions in general.

> Still, this is one more hint the flow as whole need some attention and some improvements.

> [and all the other things]

Yeah, completely agree! This definitely needs some more love. There are so many layers of things, making it all fuzzy and hard to reason about, and debug... :/ 

Maybe someone from sig-testing have more insight into what we should do, and how we should do it.

**Comment by ehashman** [comment]
/cc @BenTheElder
perhaps?

**Comment by BenTheElder** [comment]
On receiving the timeout signal the process should begin cleanup / teardown, possibly proceeded by log dumping. It should not ignore it.

We set a large grace period on prow.k8s.io to feasibly allow cleanup.

When the kill signal is recurved that is it.

If you're hitting the timeout regularly you need to increase the configured timeout or speed up the tests.

**Comment by odinuge** [comment]
> On receiving the timeout signal the process should begin cleanup / teardown, possibly proceeded by log dumping. It should not ignore it.

Yes. The current behavior is that the test runner get killed by the `SIGINT` it gets from prow (or what you would call it), and the logs will therefore not be printed and artifacts are not fetched.

With https://github.com/kubernetes/kubernetes/pull/102209, the test runner will "ignore" the SIGINT from timeout, while all its children processes will be killed. It will therefore imminently start printing the logs and fetching artifacts. The runner will therefore exit cleanly quite quickly after the initial `SIGINT`.

In a perfect world we could properly terminate the ginkgo suite, but since we get `SIGNIT` to all procs, but there is currently no way to do that.

Is that an ok solution, or do you have any other ideas.

> If you're hitting the timeout regularly you need to increase the configured timeout or speed up the tests.

Yeah, that is work in progress. The issue now is that we get no logs or artifacts when tests time out, and that is quite impractical :/

**Comment by ehashman** [comment]
FYI, we now have a `/test pull-kubernetes-node-kubelet-serial` job from https://github.com/kubernetes/test-infra/pull/22284 if we want to test fixes on PRs.

**Comment by BenTheElder** [comment]
cc @amwat this scenario is something to be aware of for kubetest -> kubetest2

**Comment by fromanirh** [comment]
I'm most likely stating the obvious, but if we exclude the eviction test, the lane seems to complete. Running locally with initial success, checking if it was a lucky run.
```
make test-e2e-node PARALLELISM=1 FOCUS=Serial SKIP=\[Flaky\]|\[Benchmark\]|\[NodeSpecialFeature:.+\]|\[NodeAlphaFeature:.+\]|\[NodeFeature:Eviction\]
```

**Comment by ehashman** [comment]
@fromanirh from the PRs above, still timed out in CI after 7 hours, I can try to submit a PR to exclude the eviction tests. Can we test those separately?

**Comment by odinuge** [comment]
https://testgrid.k8s.io/sig-node-kubelet#pr-node-kubelet-serial

@ehashman @fromanirh  :tada:

**Comment by fromanirh** [comment]
/reopen
we are not done yet

**Comment by ehashman** [comment]
/close

This issue was fixed in the 1.23 release. The job was replaced by https://testgrid.k8s.io/sig-node-containerd#cos-cgroupv2-containerd-node-e2e-serial and https://testgrid.k8s.io/sig-node-cri-o#node-kubelet-serial-crio
